File: 0003-longest-substring-without-repeating-characters/solution.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if (s.Length == 0) return 0;

        HashSet<char> set = new HashSet<char>();  // To track unique characters
        int left = 0, right = 0, maxLen = 0;
        while (right < s.Length) {
            if (!set.Contains(s[right])) {
                set.Add(s[right]);
                right++;
                maxLen = Math.Max(maxLen, right - left);
            } else {
                set.Remove(s[left]);
                left++;
            }
        }

        return maxLen;
    }
}

File: 0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if(len(s)==0):
            return 0
        d = {s[0]:1}
        length = 1
        l = 0
        r = 1
        while(r<len(s)):
            if(d.get(s[r], 0) > 0):
                length = max(length, r - l)
                while(s[l]!=s[r]):
                    del d[s[l]]
                    l+=1
                del d[s[l]]
                l+=1
            
            d[s[r]] = 1
            r+=1
        
        return max(length, r - l)
            

        

